server/download.go

```diff
@@ -0,0 +1,218 @@
+		Completed: 0,
+	}
+
+	_, downloading := inProgress.LoadOrStore(digest, fileDownload)
+	if downloading {
+		// this is another client requesting the server to download the same blob concurrently
+		return monitorDownload(ctx, mp, regOpts, fileDownload, fn)
+	}
+	resp, err := requestDownload(ctx, mp, regOpts, fileDownload)
+	if err != nil {
+		return err
+	}
+	return doDownload(ctx, fileDownload, resp, fn)
+}
+
+var downloadMu sync.Mutex // mutex to check to resume a download while monitoring
+
+// monitorDownload monitors the download progress of a blob and resumes it if it is interrupted
+func monitorDownload(ctx context.Context, mp ModelPath, regOpts *RegistryOptions, f *FileDownload, fn func(api.ProgressResponse)) error {
+	tick := time.NewTicker(time.Second)
+	for range tick.C {
+		downloadMu.Lock()
+		val, downloading := inProgress.Load(f.Digest)
+		if !downloading {
+			// check once again if the download is complete
+			if fi, _ := os.Stat(f.FilePath); fi != nil {
+				downloadMu.Unlock()
+				// successfull download while monitoring
+				fn(api.ProgressResponse{
+					Digest:    f.Digest,
+					Total:     int(fi.Size()),
+					Completed: int(fi.Size()),
+				})
+				return nil
+			}
+			// resume the download
+			resp, err := requestDownload(ctx, mp, regOpts, f)
+			if err != nil {
+				downloadMu.Unlock()
+				return fmt.Errorf("resume: %w", err)
+			}
+			inProgress.Store(f.Digest, f)
+			downloadMu.Unlock()
+			return doDownload(ctx, f, resp, fn)
+		}
+		downloadMu.Unlock()
+		f, ok := val.(*FileDownload)
+		if !ok {
+			return fmt.Errorf("invalid type for in progress download: %T", val)
+		}
+		fn(api.ProgressResponse{
+			Status:    fmt.Sprintf("downloading %s", f.Digest),
+			Digest:    f.Digest,
+			Total:     int(f.Total),
+			Completed: int(f.Completed),
+		})
+	}
+	return nil
+}
+
+var chunkSize = 1024 * 1024 // 1 MiB in bytes
+
+// requestDownload requests a blob from the registry and returns the response
+func requestDownload(ctx context.Context, mp ModelPath, regOpts *RegistryOptions, f *FileDownload) (*http.Response, error) {
+	var size int64
+
+	fi, err := os.Stat(f.FilePath + "-partial")
+	switch {
+	case errors.Is(err, os.ErrNotExist):
+		// noop, file doesn't exist so create it
+	case err != nil:
+		return nil, fmt.Errorf("stat: %w", err)
+	default:
+		size = fi.Size()
+		// Ensure the size is divisible by the chunk size by removing excess bytes
+		size -= size % int64(chunkSize)
+
+		err := os.Truncate(f.FilePath+"-partial", size)
+		if err != nil {
+			return nil, fmt.Errorf("truncate: %w", err)
+		}
+	}
+
+	url := fmt.Sprintf("%s/v2/%s/blobs/%s", mp.Registry, mp.GetNamespaceRepository(), f.Digest)
+	headers := map[string]string{
+		"Range": fmt.Sprintf("bytes=%d-", size),
+	}
+
+	resp, err := makeRequest("GET", url, headers, nil, regOpts)
+	if err != nil {
+		log.Printf("couldn't download blob: %v", err)
+		return nil, err
+	}
+	// resp MUST be closed by doDownload, which should follow this function
```

Comment:
This is error prone if the user doesn't read the code. I guess you're not closing here so the lock can released before the response is copied?

Reply:
there was a point when I needed to pass the response between functions, but I refactored this to all be one function which defers closing the resp (so no foreboding comment anymore)